### PR TITLE
Add `GET /api/core/v3/rewards/{outputId}` route

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Add `GET /api/core/v3/rewards/{outputId}` route

## Links to any relevant issues

Closes https://github.com/iotaledger/iota-sdk/issues/1051

## Type of change

- Enhancement (a non-breaking change which adds functionality)

### CODE REVIEW

File: sdk/src/client/node_api/core/routes.rs [hunk 1]
```diff
@@ -103,6 +103,18 @@ impl ClientInner {
             .await
     }
 
+    /// Returns the totally available Mana rewards of an account or delegation output decayed up to epochEnd index
+    /// provided in the response. GET /api/core/v3/rewards/{outputId}
+    pub async fn get_rewards(&self, output_id: &OutputId) -> Result<ManaRewardsResponse> {
```
Comment (thibault-martinez) [inline sdk/src/client/node_api/core/routes.rs]: Just a suggestion, for the congestion route I've slightly edited the name from the TIP to make it a bit less generic, maybe do the same? `get_mana_rewards`? `get_output_rewards`?

File: sdk/src/types/api/core/response.rs [hunk 2]
```diff
@@ -364,3 +364,22 @@ pub struct CongestionResponse {
...
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(rename_all = "camelCase")
+)]
+pub struct ManaRewardsResponse {
+    ///  The starting epoch index for which the mana rewards are returned.
+    pub epoch_start: u64, // TODO: replace with `EpochIndex`
+    ///  The ending epoch index for which the mana rewards are returned, the decay is applied up to this point
+    /// (including).
```
Comment (thibault-martinez) [inline sdk/src/types/api/core/response.rs]:
```suggestion
    ///  The ending epoch index for which the mana rewards are returned, the decay is applied up to this point included.
```
Idk, this parentheses is just weird

Comment (Thoralf-M) [reply sdk/src/types/api/core/response.rs]: Please also comment at the TIP :trollface:

File: sdk/src/client/node_api/core/routes.rs [hunk 3]
```diff
@@ -103,6 +103,18 @@ impl ClientInner {
             .await
     }
 
+    /// Returns the totally available Mana rewards of an account or delegation output decayed up to epochEnd index
+    /// provided in the response. GET /api/core/v3/rewards/{outputId}
```
Comment (thibault-martinez) [inline sdk/src/client/node_api/core/routes.rs]: GET on a new line

File: sdk/src/types/api/core/response.rs [hunk 4]
```diff
@@ -364,3 +364,21 @@ pub struct CongestionResponse {
...
+/// Response of GET /api/core/v3/rewards/{outputId}.
+/// Returns the mana rewards of an account or delegation output.
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(rename_all = "camelCase")
+)]
+pub struct ManaRewardsResponse {
+    ///  The starting epoch index for which the mana rewards are returned.
```
Comment (thibault-martinez) [inline sdk/src/types/api/core/response.rs]:
```suggestion
    /// The starting epoch index for which the mana rewards are returned.
```

File: sdk/src/types/api/core/response.rs [hunk 5]
```diff
@@ -364,3 +364,21 @@ pub struct CongestionResponse {
...
+    pub epoch_start: u64, // TODO: replace with `EpochIndex`
+    ///  The ending epoch index for which the mana rewards are returned, the decay is applied up to this point included.
```
Comment (thibault-martinez) [inline sdk/src/types/api/core/response.rs]:
```suggestion
    /// The ending epoch index for which the mana rewards are returned, the decay is applied up to this point included.
```

**Review by thibault-martinez — DISMISSED** [review]

File: sdk/src/client/node_api/core/routes.rs [hunk 6]
```diff
@@ -103,6 +103,19 @@ impl ClientInner {
             .await
     }
 
+    /// Returns the totally available Mana rewards of an account or delegation output decayed up to epochEnd index
+    /// provided in the response.
+    /// GET /api/core/v3/rewards/{outputId}
+    pub async fn get_output_mana_rewards(&self, output_id: &OutputId) -> Result<ManaRewardsResponse> {
```
Comment (abdulmth) [inline sdk/src/client/node_api/core/routes.rs:115]: > Mana rewards of an *account* or delegation output

maybe `get_mana_rewards` is more accurate?

File: sdk/src/client/node_api/core/routes.rs [hunk 7]
```diff
@@ -103,6 +103,19 @@ impl ClientInner {
             .await
     }
 
+    /// Returns the totally available Mana rewards of an account or delegation output decayed up to epochEnd index
+    /// provided in the response.
```
Comment (abdulmth) [inline sdk/src/client/node_api/core/routes.rs:109]: Maybe also include the note in the description? It can be helpful for education? 🤷🏻‍♂️

**Review by abdulmth — DISMISSED** [review]

File: sdk/src/client/node_api/core/routes.rs (hunk 6 above)
Comment (thibault-martinez) [reply sdk/src/client/node_api/core/routes.rs:115]: I think both are fine tbh, it takes an OutputId as param so having Output in the name is not that bad

File: sdk/src/client/node_api/core/routes.rs (hunk 7 above)
Comment (Thoralf-M) [reply sdk/src/client/node_api/core/routes.rs:109]: In which description? :sweat_smile:

Comment (abdulmth) [reply sdk/src/client/node_api/core/routes.rs:109]: > description: >-
        Returns the totally available Mana rewards of an account or delegation output up until the current epoch index. 
        Note that rewards for an epoch only become available at the beginning of the next epoch. 
        If the end epoch of a staking feature is equal or greater than the current epoch, the rewards response will not include the potential future rewards for those epochs.
        epochStart and epochEnd indicates the actual range for which reward value is returned and decayed for.

Comment (Thoralf-M) [reply sdk/src/client/node_api/core/routes.rs:109]: Ah like this, sure

**Review by abdulmth — DISMISSED** [review]

**Review by kwek20 — DISMISSED** [review]

**Review by thibault-martinez — DISMISSED** [review]

File: sdk/src/client/node_api/core/routes.rs [hunk 8]
```diff
@@ -103,6 +103,23 @@ impl ClientInner {
...
+    /// Note that rewards for an epoch only become available at the beginning of the next epoch. If the end epoch of a
+    /// staking feature is equal or greater than the current epoch, the rewards response will not include the potential
+    /// future rewards for those epochs. epoch_start and epoch_end indicates the actual range for which reward value is
+    /// returned and decayed for.
+    /// GET /api/core/v3/rewards/{outputId}
+    pub async fn get_output_mana_rewards(&self, output_id: &OutputId) -> Result<ManaRewardsResponse> {
```
Comment (DaughterOfMars) [inline sdk/src/client/node_api/core/routes.rs:115]: put this in a new section for rewards?

File: sdk/src/client/node_api/core/routes.rs [hunk 9]
```diff
@@ -103,6 +103,23 @@ impl ClientInner {
             .await
     }
 
+    /// Returns the totally available Mana rewards of an account or delegation output decayed up to epochEnd index
```
Comment (DaughterOfMars) [inline sdk/src/client/node_api/core/routes.rs]:
```suggestion
    /// Returns the total available Mana rewards of an account or delegation output decayed up to epochEnd index
```

Comment (DaughterOfMars) [reply sdk/src/client/node_api/core/routes.rs]: Also what is epochEnd?

File: sdk/src/client/node_api/core/routes.rs [hunk 10]
```diff
@@ -103,6 +103,23 @@ impl ClientInner {
...
+    /// provided in the response.
+    /// Note that rewards for an epoch only become available at the beginning of the next epoch. If the end epoch of a
+    /// staking feature is equal or greater than the current epoch, the rewards response will not include the potential
+    /// future rewards for those epochs. epoch_start and epoch_end indicates the actual range for which reward value is
```
Comment (DaughterOfMars) [inline sdk/src/client/node_api/core/routes.rs]: and different casing of epoch start/end here?

File: sdk/src/client/node_api/core/routes.rs (hunk 9 above)
Comment (Thoralf-M) [reply sdk/src/client/node_api/core/routes.rs]: It's the epoch until which the Mana rewards are calculated

File: sdk/src/client/node_api/core/routes.rs [hunk 11]
```diff
@@ -103,11 +103,13 @@ impl ClientInner {
...
     }
 
+    // Reward routes.
+
     /// Returns the total available Mana rewards of an account or delegation output decayed up to epochEnd index
     /// provided in the response.
     /// Note that rewards for an epoch only become available at the beginning of the next epoch. If the end epoch of a
     /// staking feature is equal or greater than the current epoch, the rewards response will not include the potential
-    /// future rewards for those epochs. epoch_start and epoch_end indicates the actual range for which reward value is
+    /// future rewards for those epochs. epochStart and epochEnd indicates the actual range for which reward value is
```
Comment (DaughterOfMars) [inline sdk/src/client/node_api/core/routes.rs]: Can you also surround these fields in `back ticks`?

**Review by thibault-martinez — DISMISSED** [review]

**Review by abdulmth — DISMISSED** [review]

File: sdk/src/client/node_api/core/routes.rs (hunk 11 above)
Comment (kwek20) [reply sdk/src/client/node_api/core/routes.rs]: also, theyre not in 🐍

Comment (Thoralf-M) [reply sdk/src/client/node_api/core/routes.rs]: @kwek20 changed this so it's aligned with the other text and TIP

File: sdk/src/client/node_api/core/routes.rs (hunk 8 above)
Comment (Thoralf-M) [reply sdk/src/client/node_api/core/routes.rs:115]: Added a comment further above if that's what you meant

**Review by abdulmth — DISMISSED** [review]

**Review by kwek20 — DISMISSED** [review]

File: sdk/src/client/node_api/core/routes.rs (hunk 8 above)
Comment (DaughterOfMars) [reply sdk/src/client/node_api/core/routes.rs:115]: I meant this is under the routes marked ```// Node routes.```

File: sdk/src/client/node_api/core/routes.rs [hunk 12]
```diff
@@ -103,6 +103,25 @@ impl ClientInner {
             .await
     }
 
+    // Reward routes.
+
+    /// Returns the total available Mana rewards of an account or delegation output decayed up to epochEnd index
```
Comment (DaughterOfMars) [inline sdk/src/client/node_api/core/routes.rs]: missed this with back ticks

Comment (Thoralf-M) [inline sdk/src/client/node_api/core/routes.rs]:
```suggestion
    /// Returns the total available Mana rewards of an account or delegation output decayed up to `epochEnd` index
```

File: sdk/src/client/node_api/core/routes.rs (hunk 8 above)
Comment (Thoralf-M) [reply sdk/src/client/node_api/core/routes.rs:115]: Yeah and further below is `// Blocks routes.`, so I added `// Reward routes.`

**Review by kwek20 — APPROVED** [review]

File: sdk/src/client/node_api/core/routes.rs (hunk 8 above)
Comment (DaughterOfMars) [reply sdk/src/client/node_api/core/routes.rs:115]: Ah, I see it now.

**Review by DaughterOfMars — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Review by abdulmth — APPROVED** [review]

**Pull request merged**